ruby/ql/lib/codeql/ruby/frameworks/ActionDispatch.qll

```diff
@@ -0,0 +1,855 @@
+     * ```
+     *
+     * the components are:
+     *
+     * | n | component
+     * |---|----------
+     * | 0 | baz
+     * | 1 | bar
+     * | 2 | foo
+     */
+    string getPathComponent(int n) {
+      if n = 0
+      then result = this.getLastPathComponent()
+      else result = this.getParentBlock().getParent(n - 1).getPathComponent()
+    }
+
+    /**
+     * Gets the full path of the route.
+     */
+    string getPath() {
+      result =
+        concat(int n |
+          this.getPathComponent(n) != ""
+        |
+          // Strip leading and trailing slashes from each path component before combining
+          stripSlashes(this.getPathComponent(n)), "/" order by n desc
+        )
+    }
+
+    /**
+     * Get a URL capture. This is a wildcard URL segment whose value is placed in `params`.
+     * For example, in
+     * ```ruby
+     * get "/foo/:bar/baz", to: "users#index"
+     * ```
+     * the capture is `:bar`.
+     * We don't currently make use of this, but it may be useful in future to more accurately
+     * model the contents of the `params` hash.
+     */
+    string getACapture() { result = this.getPathComponent(_).regexpFind(":[^:/]+", _, _) }
+  }
+
+  /**
+   * A route generated by an explicit call to `get`, `post`, etc.
+   *
+   * ```ruby
+   * get "/photos", to: "photos#index"
+   * put "/photos/:id", to: "photos#update"
+   * ```
+   */
+  private class ExplicitRoute extends Route, TExplicitRoute {
+    RouteBlock parentBlock;
+
+    ExplicitRoute() { this = TExplicitRoute(parentBlock, method) }
+
+    override string getAPrimaryQlClass() { result = "ExplicitRoute" }
+
+    override RouteBlock getParentBlock() { result = parentBlock }
+
+    override string getLastPathComponent() {
+      result = method.getArgument(0).(StringlikeLiteral).getValueText()
+    }
+
+    override string getLastControllerComponent() {
+      result = method.getKeywordArgument("controller").getValueText()
+      or
+      not exists(method.getKeywordArgument("controller")) and
+      (
+        result = extractController(this.getActionString())
+        or
+        // If not controller is specified, and we're in a `resources` route block, use the controller of that route.
+        // For example, in
+        //
+        // resources :posts do
+        //   get "timestamp", to: :timestamp
+        // end
+        //
+        // The route is GET /posts/:post_id/timestamp => posts/timestamp
+        not exists(extractController(this.getActionString())) and
+        exists(ResourcesRoute r |
+          r.getDefiningMethodCall() = parentBlock.(ResourcesRouteBlock).getDefiningMethodCall()
+        |
+          result = r.getLastControllerComponent()
+        )
+      )
+    }
+
+    private string getActionString() {
+      result = method.getKeywordArgument("to").(StringlikeLiteral).getValueText()
+      or
+      method.getKeywordArgument("to").(MethodCall).getMethodName() = "redirect" and
+      result = "<redirect>#<redirect>"
```

Comment:
Is this a placeholder value?

Reply:
Took me a while to remember what this is for! It covers cases like
```
get "/stories" => redirect("/posts")
```

In this case there's no controller/action combo that we can map `GET /stories` to, except by trying to resolve `/posts` to a route and going from there. That seemed like a lot of effort so instead we just return `<redirect>#<redirect>` for the action string.

Thinking about it again, perhaps we should just return no result in this case?

Reply:
It's probably okay as it is - by doing this it looks like `getAction()` will return `"<redirect>"`, so that's a bit more informative than having no result.